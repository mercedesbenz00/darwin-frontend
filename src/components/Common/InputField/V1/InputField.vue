<template>
  <div
    class="inputfield"
    :class="{
      'inputfield--dark': theme === 'dark',
      'inputfield--light': theme === 'light'
    }"
  >
    <div
      v-if="label"
      class="inputfield__label"
      :class="{'inputfield__label--focus': focus}"
    >
      {{ label }}
    </div>
    <div class="inputfield__wrapper">
      <input
        :id="id"
        ref="inputfield"
        v-model="inputText"
        class="inputfield__input"
        :class="{
          'inputfield__input--error': !!internalError,
          'inputfield__input--non-empty': !isEmpty,
        }"
        :type="type"
        :placeholder="placeholder"
        :autocomplete="autocomplete"
        :name="name"
        :readonly="readonly"
        :required="required"
        :min="min"
        :max="max"
        :maxlength="maxlength"
        :disabled="disabled"
        @focus="onFocus"
        @blur="onBlur"
        @keydown.stop
        @keyup.stop
        @keypress.stop
        @keydown.enter.stop="onEnter"
        @keydown.esc.stop="onEscape"
      >
      <div
        v-if="internalError"
        class="inputfield__error"
      >
        {{ internalError }}
      </div>
      <template v-else>
        <div
          v-if="hint"
          class="inputfield__hint"
        >
          {{ hint }}
        </div>
        <div class="inputfield__modifier">
          <slot name="modifier" />
        </div>
        <div class="inputfield__advisor">
          <slot name="advisor" />
        </div>
      </template>
    </div>
  </div>
</template>

<script lang="ts">
import isEmpty from 'validator/lib/isEmpty'
import { Component, Prop, Vue, Watch } from 'vue-property-decorator'

const themeValidator = (value: string) => ['dark', 'light'].indexOf(value) !== -1

@Component({ name: 'input-field' })
export default class InputField extends Vue {
  $refs!: Vue['$refs'] & {
    inputfield: HTMLInputElement
  }

  @Prop({ required: true })
  value!: string | number

  @Prop({ required: false, default: null })
  label!: string | null

  @Prop({ required: false })
  id?: string | number

  @Prop({ required: false })
  name?: string

  @Prop({ required: false, default: null })
  placeholder!: string | null

  @Prop({ required: false })
  hint?: string

  @Prop({ required: false })
  autocomplete?: string

  @Prop({ required: false })
  min?: number

  @Prop({ required: false })
  max?: number

  @Prop({ required: false })
  required?: string

  @Prop({ required: false })
  maxlength?: number

  @Prop({ required: false, default: 'text' })
  type!: string

  @Prop({ required: false, default: false, type: Boolean })
  autoFocus!: boolean

  @Prop({ required: false, default: false, type: Boolean })
  autofill!: boolean

  @Prop({ required: false, default: false, type: Boolean })
  readonly!: boolean

  @Prop({ required: false, default: false, type: Boolean })
  disabled!: boolean

  /**
   * Color theming for the field.
   *
   * Default is 'dark' and should be used on non-white backgrounds.
   * On white backgrounds, the theme should be 'light
   */
  @Prop({ required: false, default: 'dark', validator: themeValidator })
  theme!: 'dark' | 'light'

  // validation error logic

  /**
   * Allows setting error the conventional way, as a prop
   */
  @Prop({ required: false, default: null })
  error!: string | null

  /**
   * For usage with old, unconventional and difficult to test `setError`
   */
  internalError: string | null = null

  public setError (error?: string) {
    this.internalError = error || null
  }

  @Watch('inputText')
  onInputText () {
    this.internalError = null
    this.$refs.inputfield.value = this.inputText.toString()
  }

  @Watch('error')
  onError (value: string) {
    this.internalError = value
  }

  focus: boolean = false

  get inputText (): string | number {
    return this.value
  }

  set inputText (val: string | number) {
    this.$emit('input', val)
    this.$emit('change', val)
  }

  get isEmpty (): boolean {
    return !this.inputText || isEmpty(`${this.inputText}`)
  }

  mounted () {
    if (this.autoFocus) {
      this.$nextTick(() => {
        this.setFocus()
      })
    }

    if (this.error) {
      this.internalError = this.error
    }
  }

  public setFocus (focus: boolean = true) {
    const inputfield = this.$refs.inputfield
    if (!inputfield) { return }
    setTimeout(() => {
      focus
        ? inputfield.focus()
        : inputfield.blur()
    }, 500)
  }

  onFocus (): void {
    this.focus = true
  }

  onBlur (): void {
    this.$emit('blur')
    setTimeout(() => {
      this.focus = false
    }, 300)
  }

  onEnter (e: KeyboardEvent): void {
    this.$emit('enter', e)
  }

  onEscape (e: KeyboardEvent): void {
    this.$emit('escape', e)
  }
}
</script>

<style lang="scss" scoped>
$inputHeight: 36px;
$fontSize: 14px;

.inputfield {
  position: relative;
  @include col;
  width: 100%;
}

.inputfield__input {
  transition: background-color .2s ease;
  transition: color .2s ease;
  color: $color90Black;
  width: 100%;
  height: $inputHeight;
  border: none;
  border-radius: $border-radius-default;
  font-size: $fontSize;
  font-family: $fontFamilyDefault;
  padding: 6px 9px;
  background: $colorAliceBlue;
  box-shadow: inset 0px 2px 4px rgba(145, 169, 192, 0.3);

  &:focus, &:active {
    border: none;
    outline: none !important;
  }

  &:disabled {
    color: $colorAliceNight;
    background: $colorLineGrey;
  }

  &::placeholder {
    color: $colorAliceNight;
  }
}

// this seems to be the common style now used across the app
// we should consider making it default
// basic rules are, different fg and bg when enabled/disabled
.inputfield--dark {
  .inputfield__input {
    background: $colorAliceShade;
  }

  .inputfield__input.inputfield__input--non-empty:not(:disabled),
  .inputfield__input:focus:not(:disabled) {
    background: $colorAliceShade;
  }

  .inputfield__input:disabled {
    background: $colorAliceShadow;
    color: $colorAliceNight;
    box-shadow: none;
  }
}

.inputfield--light {
  .inputfield__input {
    background: $colorAliceBlue;
  }
}

.inputfield__label {
  @include typography(md-1, default);
  color: $colorAliceNight;
  margin-bottom: 7px;

  &--focus {
    color: $colorPrimaryDark;
  }
}

.inputfield__wrapper {
  @include row--center;
  width: 100%;
  position: relative;
}

.inputfield__input--error {
  margin-top: 0;
  color: $colorPink;
}

.inputfield__hint {
  position: absolute;
  left: 0;
  bottom: -19px;
  @include typography(md, default);
  color: $colorAliceNight;
}

.inputfield__modifier {
  position: absolute;
  right: 0;
  top: 0;
  bottom: 0;
  max-width: 120px;
  height: 100%;
  align-items: center;
  @include row;
  justify-content: flex-end;
}

.inputfield__advisor {
  position: absolute;
  right: 0;
  bottom: -17px;
  @include row;
  justify-content: flex-end;
}

.inputfield__error {
  position: absolute;
  left: 0;
  bottom: -19px;
  @include typography(md, default, bold);
  color: $colorPink;
}
</style>
