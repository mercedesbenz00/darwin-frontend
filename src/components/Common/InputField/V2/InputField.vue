<template>
  <component
    class="inputfield_wrapper"
    :is="tag"
    tabindex="1"
    :v-bind="$attrs"
    :v-on="$listeners"
  >
    <div
      id="InputFieldScrollable"
      class="inputfield__container"
      :class="{
        focused: focus,
        disabled: disabled,
        error: !!hasError,
        warning: !!hasWarning,
        'inputfield__container--wrap': shouldWrap
      }"
    >
      <div
        v-if="$slots['left-icon']"
        class="inputfield-icon__wrapper"
        @click="input.focus()"
      >
        <slot name="left-icon" />
      </div>
      <div
        v-if="multiple"
        class="inputfield-items__wrapper"
      >
        <badge
          v-for="({ id, label, color }) in (items || [])"
          :key="id"
          :label="label"
          :color="color"
          size="medium"
          deletable
          wrap
          no-capitalize
          @delete="$emit('delete', id)"
        />
        <input
          :id="id"
          class="inputfield inputfield--multiple"
          :class="{ 'inputfield--multiple--wrap': shouldWrap}"
          :placeholder="placeholder"
          v-bind="$props"
          v-model="inputText"
          ref="input"
          @focus="onFocus"
          @blur="onBlur"
          @keypress="isNumber($event)"
          @keydown.enter.stop="onEnter"
          @keydown.esc.stop="onEscape"
        >
      </div>
      <input
        v-else
        :id="id"
        class="inputfield"
        :placeholder="placeholder"
        v-bind="$props"
        v-model="inputText"
        ref="input"
        @change="onFocus"
        @focus="onFocus"
        @blur="onBlur"
        @keypress="isNumber($event)"
        @keydown.enter.stop="onEnter"
        @keydown.esc.stop="onEscape"
      >
      <div
        class="inputfield-icon__wrapper"
        v-if="$slots['right-icon']"
      >
        <slot name="right-icon" />
      </div>
      <div
        v-if="hasError"
        class="error-icon__wrapper"
      >
        <icon-duotone-warn />
      </div>
    </div>
    <div
      v-if="error"
      class="inputfield__error"
    >
      {{ error }}
    </div>
    <div
      v-if="warning"
      class="inputfield__warning"
    >
      {{ warning }}
    </div>
    <div
      v-if="info"
      class="inputfield__info"
    >
      {{ info }}
    </div>
  </component>
</template>

<script lang="ts">
import { computed, defineComponent, ref } from 'vue'

import { IconDuotoneWarn } from '@/assets/icons/V2/Duotone'
import { Badge } from '@/components/Common/Badge'

import { InputFieldProps } from './types'

/**
 * Generic input field component to be used across the board.
 *
 * Wraps an input element with a couple of extra elements and slots, to support
 * giving it extra content on the left and right side, an error, info or warning message, etc.
 */
export default defineComponent({
  name: 'InputField',
  components: { IconDuotoneWarn, Badge },
  props: {
    /**
     * The actual value of the input element.
     * Emits a change event when it changes, so it can be used with v-model.
     */
    value: {
      required: true,
      type: String as () => InputFieldProps['value']
    },
    /**
     * Optional tag attribute which is bound to the input element
     */
    tag: {
      required: false,
      default: 'div',
      type: String as () => InputFieldProps['tag']
    },
    /**
     * Optional id attribute which is bound to the input element
     */
    id: {
      required: false,
      default: null,
      type: String as () => InputFieldProps['id']
    },
    /**
     * Optional placeholder attribute for the input element
     */
    placeholder: {
      required: false,
      default: null,
      type: String as () => InputFieldProps['placeholder']
    },
    /**
     * For numeric inputs, optional value for the `min` attribute
     */
    min: {
      required: false,
      default: null,
      type: Number as () => InputFieldProps['min']
    },
    /**
     * For numeric inputs, optional value for the `max` attribute
     */
    max: {
      required: false,
      default: null,
      type: Number as () => InputFieldProps['max']
    },
    /**
     * Binds to the `required` attribute of the input field.
     */
    required: {
      required: false,
      default: false,
      type: Boolean as () => InputFieldProps['required']
    },
    /**
     * For mainly text inputs, optional value for the html `max-length` attribute
     */
    maxLength: {
      required: false,
      default: null,
      type: Number as () => InputFieldProps['maxLength']
    },
    /**
     * Optional value for the type attribute of the input element.
     * Defaults to 'text'.
     */
    type: {
      required: false,
      default: 'text',
      type: String as () => InputFieldProps['type']
    },
    /**
     * Binds to the 'disabled' html attribute of the input element
     */
    disabled: {
      required: false,
      default: false,
      type: Boolean as () => InputFieldProps['disabled']
    },
    /**
     * Optional value for the error message, rendered below the input field.
     */
    error: {
      required: false,
      default: null,
      type: String as () => InputFieldProps['error']
    },
    /**
     * Optional value for the info message, rendered below the input field.
     */
    info: {
      required: false,
      default: null,
      type: String as () => InputFieldProps['info']
    },
    /**
     * Optional value for the warning message, rendered below the input field.
     */
    warning: {
      required: false,
      default: null,
      type: String as () => InputFieldProps['warning']
    },
    /**
     * Focuses the input element automatically on mount if true.
     * With multiple autofocus components on the same page, the last one mounted wins.
     */
    autofocus: {
      required: false,
      default: false,
      type: Boolean as () => InputFieldProps['autofocus']
    },
    /**
     * Keep an array of values rendered as badges
     */
    items: {
      required: false,
      default: () => [],
      type: Array as () => InputFieldProps['items']
    },
    /**
     * Keep an array of values rendered as badges
     */
    multiple: {
      required: false,
      default: false,
      type: Boolean as () => InputFieldProps['multiple']
    },
    /**
     * used togheter with multiple its height adapt to
     * the contained items
     */
    wrap: {
      required: false,
      default: false,
      type: Boolean as () => InputFieldProps['wrap']
    },
    /**
     * Text input but allows only numbers as input
     */
    largeNumber: {
      required: false,
      default: false,
      type: Boolean as () => InputFieldProps['largeNumber']
    }
  },
  mounted () {
    // ensure we focus the input field if autofocus is enabled
    if (!!this.input && this.autofocus) { this.input.focus() }
  },
  setup (props, { emit }) {
    // toggles between true/false on input field focus/blur
    const focus = ref(false)

    /**
     * Check if input should wrap
     */
    const shouldWrap = computed((): boolean => {
      return !!props.multiple && !!props.wrap && (props.items || []).length > 0
    })

    /**
     * Delegates the value prop on get, but emits events on set,
     * to support v-model and syncing with errors/warnings
     */
    const inputText = computed({
      get: () => props.value,
      set: (val) => {
        // clear error
        if (props.error) { emit('update:error', null) }
        // clear warning
        if (props.warning) { emit('update:warning', null) }
        emit('input', val)
        emit('change', val)
      }
    })

    /**
     * Delegates the items prop on get, but emits events on set,
     * to support v-model and syncing with errors/warnings
     */
    const inputItems = computed({
      get: () => props.items,
      set: (val) => emit('change:multiple', val)
    })

    /**
     * Boolean returning true if the error prop is set.
     * Means the error will be rendered.
     */
    const hasError = computed(() => !!props.error)

    /**
     * Boolean returning true if the warning prop is set.
     * Means the warning will be rendered.
     */
    const hasWarning = computed(() => !!props.warning)

    /**
     * Event handler called when the input element receives focus
     */
    const onFocus = (): void => {
      focus.value = true
      emit('focus')
    }

    /**
     * Event handler called when the blur element is blurred
     */
    const onBlur = (): void => {
      focus.value = false
      emit('blur')
    }
    /**
     * Holds element reference to the input field.
     * We use it in the mounted hook to achieve autofocus behavior.
     */
    const input = ref<HTMLInputElement | null>(null)

    /**
     * Event handler called when the enter key event is received by the input field,
     * meaning user pressed enter while the input was focused.
     */
    const onEnter = (event: KeyboardEvent): void => {
      if (props.multiple) {
        inputItems.value = [
          ...(inputItems.value || []),
          { label: `${inputText.value}` }
        ]
        inputText.value = ''
      }
      emit('enter', { event, value: `${input.value?.value}`, input: input.value })
    }

    const onEscape = (e: KeyboardEvent): void => {
      emit('escape', e)
    }

    const isNumber = (evt: KeyboardEvent): boolean | void => {
      if (!props.largeNumber) { return true }
      evt = evt || window.event
      const charCode = evt.which ? evt.which : evt.keyCode
      if (charCode > 31 && (charCode < 48 || charCode > 57) && charCode !== 46) {
        evt.preventDefault()
      } else {
        return true
      }
    }

    const setFocus = (): void => {
      if (input.value) {
        input.value.focus()
      }
    }

    return {
      props,
      focus,
      input,
      shouldWrap,
      inputText,
      inputItems,
      hasError,
      hasWarning,
      onFocus,
      onBlur,
      setFocus,
      isNumber,
      onEnter,
      onEscape
    }
  }
})

</script>

<style lang="scss" scoped>
.inputfield_wrapper {
  display: block;
  width: 100%;
}

.inputfield__container {
  transition: border 225ms ease, background-color 225ms ease;

  display: flex;
  align-items: center;
  justify-content: space-between;
  height: 36px;
  width: 100%;
  border-radius: 10px;
  border: 1px solid $colorNeutralsLight300;
  background: $colorNeutralsLightWhite;
  padding: 0 8px;

  &--wrap {
    flex-wrap: wrap;
    height: auto !important;
    max-height: 240px;
    padding: 8px;
    gap: 8px;
    @include scrollbarV2;

    &:focus,
    &:active {
      outline: none;
    }
  }

  &:hover {
    background: $colorNeutralsLight100;
  }

  &--focused {
    border: 1px solid $colorInteractivePrimaryDefault;
    background-color: $colorSurfaceDefault;
    caret-color: $colorInteractivePrimaryDefault;
  }

  &--disabled {
    background-color: $colorInteractiveSecondaryDisabled;
    cursor: not-allowed;
  }

  &--error {
    border: 1.5px solid $colorStatusNegative;
    border: none;
  }

  &--warning {
    border: 1.5px solid $colorStatusWarning;
    border: none;
  }
}

.inputfield {
  @include typography(md-1, inter, 500);
  color: $colorNeutralsLight900;
  background-color: transparent;
  height: 100%;
  width: 100%;

  &--multiple--wrap {
    display: flex;
    flex: 1 0;
    height: 22px;
    @include ellipsis(1, md-1);
  }

  &:disabled {
    color: $colorNeutralsLight400;
    cursor: not-allowed;

    &::placeholder {
      color: $colorNeutralsLight400;
    }
  }

  &::placeholder {
    color: $colorNeutralsLight600;
  }
}

.error-icon__wrapper {
  display: flex;
  justify-content: center;
  align-items: center;
  width: 20px;
  height: 20px;
  margin-right: 12px;
}

.inputfield-items__wrapper {
  @include row;
  flex-wrap: wrap;
  gap: 4px;
  width: auto;
  height: 100%;
  width: 100%;

  .badge {
    max-width: 160px;

    &:deep(.badge__content__label) {
      @include ellipsis(1, md);
    }
  }
}

.inputfield-icon__wrapper {
  display: flex;
  justify-content: center;
  align-items: center;
  width: 20px;
  height: 20px;

  &:first-child {
    margin-right: 4px;
  }
}

.inputfield__error {
  margin-top: 4px;
  @include typography(sm, inter, 500);
  color: $colorStatusNegative;
}

.inputfield__warning {
  margin-top: 4px;
  @include typography(sm, inter, 500);
  color: $colorStatusWarning;
}
.inputfield__info {
  margin-top: 4px;
  @include typography(sm, inter, 500);
}
</style>
