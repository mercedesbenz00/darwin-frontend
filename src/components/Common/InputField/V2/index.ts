export { default as InputField } from './InputField.vue'
export * from './types'
