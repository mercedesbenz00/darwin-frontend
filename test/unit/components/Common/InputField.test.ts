import { shallowMount, createLocalVue } from '@vue/test-utils'

import InputField from '@/components/Common/InputField/V1/InputField.vue'

const localVue = createLocalVue()

let propsData: {
  value: string | number

  autocomplete?: string
  autofill?: boolean
  autoFocus?: boolean
  disabled?: boolean
  error?: string
  feeds?: { id: string | number, value: string | number }[]
  hint?: string
  id?: string | number
  label?: string
  max?: number
  maxlength?: number
  min?: number
  name?: string
  placeholder?: string
  readonly?: boolean
  required?: string
  theme?: 'dark' | 'light'
  type?: string
}

describe('basic behavior', () => {
  beforeEach(() => {
    propsData = { value: 'Foo' }
  })

  it('matches snapshot', () => {
    const wrapper = shallowMount(InputField, { localVue, propsData })
    expect(wrapper).toMatchSnapshot()
  })

  it('binds value', () => {
    const wrapper = shallowMount(InputField, { localVue, propsData })
    const input = (wrapper.find('input').element as HTMLInputElement)
    expect(input.value).toEqual('Foo')
  })

  it('emits change/input', () => {
    const wrapper = shallowMount(InputField, { localVue, propsData })
    expect(wrapper.find('input').setValue('Bar'))
    expect(wrapper.emitted().input).toEqual([['Bar']])
    expect(wrapper.emitted().change).toEqual([['Bar']])
  })
})

describe('prop-based validation error', () => {
  propsData = { value: 'Foo' }

  it('matches snapshot', async () => {
    const wrapper = shallowMount(InputField, { localVue, propsData })
    wrapper.setProps({ error: 'Foo Error' })
    await wrapper.vm.$nextTick()
    expect(wrapper).toMatchSnapshot()
  })

  it('renders error', async () => {
    const wrapper = shallowMount(InputField, { localVue, propsData })
    wrapper.setProps({ error: 'Foo Error' })
    await wrapper.vm.$nextTick()
    expect(wrapper.find('.inputfield__error').text()).toEqual('Foo Error')
  })

  it('clears error when typing', async () => {
    const wrapper = shallowMount(InputField, { localVue, propsData })
    wrapper.setProps({ error: 'Foo Error' })
    await wrapper.vm.$nextTick()
    expect(wrapper.find('.inputfield__error').exists()).toBe(true)

    wrapper.setProps({ value: 'Foo2' })
    await wrapper.vm.$nextTick()
    expect(wrapper.find('.inputfield__error').exists()).toBe(false)

    wrapper.setProps({ error: 'Foo Error 2' })
    await wrapper.vm.$nextTick()
    expect(wrapper.find('.inputfield__error').exists()).toBe(true)
  })
})

describe('setError-based validation error', () => {
  propsData = { value: 'Foo' }

  it('matches snapshot', async () => {
    const wrapper = shallowMount(InputField, { localVue, propsData })

    const component = wrapper.vm as any
    component.setError('Foo Error')
    await wrapper.vm.$nextTick()
    expect(wrapper).toMatchSnapshot()
  })

  it('renders error', async () => {
    const wrapper = shallowMount(InputField, { localVue, propsData })
    const component = wrapper.vm as any

    component.setError('Foo Error')
    await wrapper.vm.$nextTick()
    expect(wrapper.find('.inputfield__error').text()).toEqual('Foo Error')
  })

  it('clears error when typing', async () => {
    const wrapper = shallowMount(InputField, { localVue, propsData })
    const component = wrapper.vm as any

    component.setError('Foo Error')
    await wrapper.vm.$nextTick()
    expect(wrapper.find('.inputfield__error').exists()).toBe(true)

    wrapper.setProps({ value: 'Foo2' })
    await wrapper.vm.$nextTick()
    expect(wrapper.find('.inputfield__error').exists()).toBe(false)

    component.setError('Foo Error 2')
    await wrapper.vm.$nextTick()
    expect(wrapper.find('.inputfield__error').exists()).toBe(true)
  })
})
