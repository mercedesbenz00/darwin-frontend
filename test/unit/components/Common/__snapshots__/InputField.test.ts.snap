// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic behavior matches snapshot 1`] = `
<div class="inputfield inputfield--dark">
  <!---->
  <div class="inputfield__wrapper"><input type="text" class="inputfield__input inputfield__input--non-empty">
    <!---->
    <div class="inputfield__modifier"></div>
    <div class="inputfield__advisor"></div>
  </div>
</div>
`;

exports[`prop-based validation error matches snapshot 1`] = `
<div class="inputfield inputfield--dark">
  <!---->
  <div class="inputfield__wrapper"><input type="text" class="inputfield__input inputfield__input--error inputfield__input--non-empty">
    <div class="inputfield__error">
      Foo Error
    </div>
  </div>
</div>
`;

exports[`setError-based validation error matches snapshot 1`] = `
<div class="inputfield inputfield--dark">
  <!---->
  <div class="inputfield__wrapper"><input type="text" class="inputfield__input inputfield__input--error inputfield__input--non-empty">
    <div class="inputfield__error">
      Foo Error
    </div>
  </div>
</div>
`;
